man/convert2da.1
.Dd July 23, 2015
.Dt CONVERT2DA 1
.Os
.Sh NAME
.Nm convert2da
.Nd BioWare 2DA/GDA to 2DA/CSV converter
.Sh SYNOPSIS
.Nm convert2da
.Op Ar options
.Ar
.Sh DESCRIPTION
.Nm
converts BioWare's 2DA and GDA files into (cleanly formatted)
ASCII 2DA or CSV files.
2DA files are two-dimensional arrays, or tables, holding various
information and used throughout many of the BioWare games.
They can come in two flavors: binary and plain-text ASCII.
GDA files are conceptually the same, but they
store their data inside a GFF, a binary hierarchical format.
Unlike 2DA files, a GDA does not store its column header name
directory.
Instead, only a CRC hash of the name encoded in
little-endian UTF-16 is stored.
.Pp
This tool can read all of those formats and produces an easily
readable, cleanly formatted ASCII 2DA.
Alternatively, it can create a CSV file for further processing.
It also contains a lookup table to convert GDA column header hashes
back to readable names.
Not all column header names are known, though.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl o Ar file
.It Fl Fl output Ar file
Write the output to this file.
If this option is not used, the output is written to
.Dv stdout .
.It Fl a
.It Fl Fl 2da
Convert the 2DA or GDA file into an ASCII 2DA file.
This is the default mode of operation.
.It Fl b
.It Fl Fl 2dab
Convert the 2DA or GDA file into a binary 2DA file.
.It Fl c
.It Fl Fl csv
Convert the 2DA or GDA file into an CSV file.
.El
.Bl -tag -width xx -compact
.It Ar file
The name of the 2DA or GDA file to read.
.Pp
If more than one input file is given, they must all be GDA files
and use the same column layout. They will be pasted together and
converted as one GDA. This mimics how the M2DA (multiple 2DA/GDA)
work in the
.Em Dragon Age
games.
.El
.Sh EXAMPLES
Convert the 2DA file1.2da into an ASCII 2DA
.Pa file2.2da :
.Pp
.Dl $ convert2da -a -o file2.2da file1.2da
.Pp
Convert the 2DA file1.2da into an ASCII 2DA on
.Dv stdout :
.Pp
.Dl $ convert2da -a file1.2da
.Pp
Convert the 2DA
.Pa file1.2da
into a binary 2DA
.Pa file2.2da :
.Pp
.Dl $ convert2da -b -o file2.2da file1.2da
.Pp
Convert the 2DA
.Pa file1.2da
into a CSV file:
.Pp
.Dl $ convert2da -c -o file2.csv file1.2da
.Sh SEE ALSO
.Xr gff2xml 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/convert2da.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to convert 2DA/GDA files to 2DA/CSV.
 */

#include <cstring>
#include <cstdio>

#include <memory>

#include "src/version/version.h"

#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/stdoutstream.h"
#include "src/common/encoding.h"
#include "src/common/platform.h"
#include "src/common/cli.h"

#include "src/aurora/aurorafile.h"
#include "src/aurora/2dafile.h"
#include "src/aurora/gdafile.h"

#include "src/util.h"

enum Format {
	kFormat2DA,
	kFormat2DAb,
	kFormatCSV
};

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      std::vector<Common::UString> &files, Common::UString &outFile, Format &format);

void write2DA(Aurora::TwoDAFile &twoDA, Format format);

Aurora::TwoDAFile *get2DAGDA(Common::SeekableReadStream *stream);
void convert2DA(const Common::UString &file, const Common::UString &outFile, Format format);
void convert2DA(const std::vector<Common::UString> &files, const Common::UString &outFile, Format format);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Format format = kFormat2DA;

		int returnValue = 1;
		std::vector<Common::UString> files;
		Common::UString outFile;

		if (!parseCommandLine(args, returnValue, files, outFile, format))
			return returnValue;

		convert2DA(files, outFile, format);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      std::vector<Common::UString> &files, Common::UString &outFile,
                      Format &format) {
	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::Callback;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;

	NoOption filesOpt(false, new ValGetter<std::vector<Common::UString> &>(files, "files[...]"));
	Parser parser(argv[0], "BioWare 2DA/GDA to 2DA/CSV converter\n",
	              "If several files are given, they must all be GDA and use the same\n"
	              "column layout. They will be pasted together and printed as one GDA.\n\n"
	              "If no output file is given, the output is written to stdout.",
	              returnValue,
	              makeEndArgs(&filesOpt));

	parser.addSpace();
	parser.addOption("output", 'o', "Write the output to this file",
	                 kContinueParsing,
	                 new ValGetter<Common::UString &>(outFile, "file"));
	parser.addSpace();
	parser.addOption("2da", "Convert to ASCII 2DA (default)",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Format>(kFormat2DA,
	                 format)));
	parser.addOption("2dab", "Convert to binary 2DA", kContinueParsing,
	                 makeAssigners(new ValAssigner<Format>(kFormat2DAb,
	                 format)));
	parser.addOption("csv", "Convert to CSV", kContinueParsing,
	                 makeAssigners(new ValAssigner<Format>(kFormatCSV,
	                 format)));
	return parser.process(argv);
}

static const uint32_t k2DAID     = MKTAG('2', 'D', 'A', ' ');
static const uint32_t k2DAIDTab  = MKTAG('2', 'D', 'A', '\t');
static const uint32_t kGFFID     = MKTAG('G', 'F', 'F', ' ');

void write2DA(Aurora::TwoDAFile &twoDA, const Common::UString &outFile, Format format) {
	std::unique_ptr<Common::WriteStream> out(openFileOrStdOut(outFile));

	if      (format == kFormat2DA)
		twoDA.writeASCII(*out);
	else if (format == kFormat2DAb)
		twoDA.writeBinary(*out);
	else
		twoDA.writeCSV(*out);

	out->flush();
}

Aurora::TwoDAFile *get2DAGDA(Common::SeekableReadStream *stream) {
	std::unique_ptr<Common::SeekableReadStream> fStream(stream);

	const uint32_t id = Aurora::AuroraFile::readHeaderID(*fStream);
	fStream->seek(0);

	if ((id == k2DAID) || (id == k2DAIDTab))
		return new Aurora::TwoDAFile(*fStream);

	if (id == kGFFID) {
		Aurora::GDAFile gda(fStream.release());

		return new Aurora::TwoDAFile(gda);
	}

	throw Common::Exception("Not a 2DA or GDA file");
}

void convert2DA(const Common::UString &file, const Common::UString &outFile, Format format) {
	std::unique_ptr<Aurora::TwoDAFile> twoDA(get2DAGDA(new Common::ReadFile(file)));

	write2DA(*twoDA, outFile, format);
}

void convert2DA(const std::vector<Common::UString> &files, const Common::UString &outFile, Format format) {
	if (files.size() == 1) {
		convert2DA(files[0], outFile, format);
		return;
	}

	Aurora::GDAFile gda(new Common::ReadFile(files[0]));

	for (size_t i = 1; i < files.size(); i++)
		gda.add(new Common::ReadFile(files[i]));

	Aurora::TwoDAFile twoDA(gda);

	write2DA(twoDA, outFile, format);
}
